Mathlib/Data/List/Destutter.lean

```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
   | _ :: l => ⟨fun h => absurd h <| l.destutter'_ne_nil R, fun h => nomatch h⟩
 #align list.destutter_eq_nil List.destutter_eq_nil
 
+/-- For a relation-preserving map, `destutter'` commutes with `map`. -/
+theorem destutter_map_iff (f : α → β) (h : ∀ a b, R a b ↔ R₂ (f a) (f b)) :
+    (l.destutter R).map f = (l.map f).destutter R₂ := by
+  cases h2 : l with
+  | nil => simp -- l = []
+  | cons a as =>
+    clear h2
+    induction as generalizing a with
+    | nil => simp -- l = a :: []
+    | cons a2 bs ih => -- l = a :: a2 :: bs
+      repeat rw [map_cons, destutter_cons_cons]
+      simp_rw [← h a a2]
+      by_cases hr : (R a a2) <;>
+        simp [hr, ← destutter_cons', ih]
+
+/-- For a injective function `f`, `destutter' (·≠·)` commutes with `map f`. -/
+theorem map_destutter_ne {f : α → β} (h : Injective f) [DecidableEq α] [DecidableEq β] :
+    (l.destutter (·≠·)).map f = (l.map f).destutter (·≠·) :=
+  destutter_map_iff l _ f (fun _ _ ↦ (Injective.ne_iff h).symm)
+
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, gives the same length if
+    the first elements are not related (¬Rα a b). --/
+theorem length_destutter' [IsEquiv α Rᶜ] (hab : ¬R a b) :
+    (List.destutter' R a l).length = (List.destutter' R b l).length := by
+  induction l with
+  | nil => simp
+  | cons c cs ih =>
+    obtain hac | hac : R a c ∨ Rᶜ a c := em _
+    · have hbc : ¬Rᶜ b c := mt (_root_.trans hab) (not_not.2 hac)
+      simp [destutter', if_pos hac, if_pos (not_not.1 hbc)]
+    · have hbc : ¬R b c := trans (symm hab) hac
+      simpa [destutter', if_neg hac, if_neg hbc]
+
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, has length
+    monotonic under List.cons --/
+theorem le_length_destutter' [IsEquiv α Rᶜ] :
+    (List.destutter' R b l).length ≤ (List.destutter' R a (b :: l)).length := by
```

Comment:
Can you generalise that to
```suggestion
theorem Suffix.length_destutter'_mono [IsEquiv α Rᶜ] (h : a :: l₁ <:+ b :: l₂) :
    (List.destutter' R a l₁).length ≤ (List.destutter' R b l₂).length := by
```

Reply:
I think that's a correct statement as written, but it would probably just be an induction with `le_length_destutter'` as the inductive step. I don't really know anything about `Suffix` and I'm not super motivated to dive into that atm, so I hope it's alright I skip this one.

Reply:
Mind leaving 
```suggestion
/-
TODO: Replace this lemma by the more general version:
theorem Suffix.length_destutter'_mono [IsEquiv α Rᶜ] (h : a :: l₁ <:+ b :: l₂) :
    (List.destutter' R a l₁).length ≤ (List.destutter' R b l₂).length
-/
theorem le_length_destutter' [IsEquiv α Rᶜ] :
    (List.destutter' R b l).length ≤ (List.destutter' R a (b :: l)).length := by
```